src/accountsdb/download.zig

```diff
@@ -295,81 +297,91 @@ pub fn downloadSnapshotsFromGossip(
 }
 
 const DownloadProgress = struct {
-    mmap: []align(std.mem.page_size) u8,
-    download_size: usize,
+    file: std.fs.File,
     min_mb_per_second: ?usize,
     logger: Logger,
 
     mb_timer: std.time.Timer,
-    bytes_read: usize = 0,
-    file_memory_index: usize = 0,
+    bytes_read: u64 = 0,
+    total_read: u64 = 0,
     has_checked_speed: bool = false,
 
     const Self = @This();
 
-    pub fn init(
+    fn init(
         logger: Logger,
         output_dir: std.fs.Dir,
         filename: []const u8,
         download_size: usize,
         min_mb_per_second: ?usize,
     ) !Self {
-        const file = try output_dir.createFile(filename, .{ .read = true });
-        defer file.close();
+        const file = try output_dir.createFile(filename, .{});
 
         // resize the file
         try file.seekTo(download_size - 1);
         _ = try file.write(&[_]u8{1});
         try file.seekTo(0);
 
-        const file_memory = try std.posix.mmap(
-            null,
-            download_size,
-            std.posix.PROT.READ | std.posix.PROT.WRITE,
-            std.posix.MAP{ .TYPE = .SHARED },
-            file.handle,
-            0,
-        );
-
         return .{
             .logger = logger,
-            .mmap = file_memory,
-            .download_size = download_size,
+            .file = file,
             .min_mb_per_second = min_mb_per_second,
-            .mb_timer = try std.time.Timer.start(),
+            .mb_timer = undefined,
         };
     }
 
-    pub fn deinit(self: *Self) void {
-        std.posix.munmap(self.mmap);
+    fn deinit(self: *Self) void {
+        self.file.close();
     }
 
-    pub fn bufferWriteCallback(ptr: [*c]c_char, size: c_uint, nmemb: c_uint, user_data: *anyopaque) callconv(.C) c_uint {
+    fn writeCallback(
+        ptr: ?[*:0]c_char,
+        size: c_uint,
+        nmemb: c_uint,
+        user_data: *anyopaque,
+    ) callconv(.C) c_uint {
+        assert(size == 1); // size will always be 1
         const len = size * nmemb;
-        var self: *Self = @alignCast(@ptrCast(user_data));
-        var typed_data: [*]u8 = @ptrCast(ptr);
+        const self: *Self = @alignCast(@ptrCast(user_data));
+        var typed_data: [*]u8 = @ptrCast(ptr.?);
         const buf = typed_data[0..len];
 
-        @memcpy(self.mmap[self.file_memory_index..][0..len], buf);
-        self.file_memory_index += len;
+        self.file.writeAll(buf) catch |err|
+            std.debug.panic("failed to write to file: {s}", .{@errorName(err)});
```

Comment:
It would probably be better to return an error instead of panicking. Seems doable.

> Your callback should return the number of bytes actually taken care of. If that amount differs from the amount passed to your callback function, it signals an error condition to the library. This causes the transfer to get aborted and the libcurl function used returns CURLE_WRITE_ERROR. 
>
> You can also abort the transfer by returning CURL_WRITEFUNC_ERROR (added in 7.87.0), which makes CURLE_WRITE_ERROR get returned.

Reply:
696c4fd0bac5fd373a70ce2c99ae52db76f2d775